src/xb-arena.c

```diff
@@ -0,0 +1,137 @@
+    if(len > XB_ARENA_CHUNKSIZE){
+        uint8_t* chunk = malloc(len);
+        xb_arena_add_chunk(arena, chunk);
+        /* No need to invalidate tail or current_chunk_end, there's still space
+        in the current chunk for more allocations */
+        return chunk;
+    }
+    /* Round allocations up to nearest whole machine word
+     * If we didn't do this, allocated structs may cost extra
+     * to access because of unaligned load/stores and may
+     * straddle multiple cache lines */
+    len = (len+sizeof(ssize_t)-1) & ~(sizeof(ssize_t)-1);
+
+    /* Allocate another chunk if the the requested size won't fit in the current one.
+     * This will always be hit for the first allocation on a new arena
+     * because tail and current_chunk_end are initialized to zero. */
+    if(arena->tail + len > arena->current_chunk_end) {        
+        uint8_t* chunk = malloc(XB_ARENA_CHUNKSIZE);
+        xb_arena_add_chunk(arena, chunk);
+        arena->tail = chunk;
+        arena->current_chunk_end = chunk + (XB_ARENA_CHUNKSIZE);
+    }
+
+    uint8_t* alloc = arena->tail;
+    arena->tail += len;
+    return (void*)alloc;
+}
+
+void xb_arena_free(XbArena* arena) {
+    for (uint32_t i=0; i < arena->n_chunks; i++) {
+        free(arena->chunks[i]);
+    }
+    free((void*)arena->chunks);
+    arena->n_chunks = 0;
+    arena->chunks_cap = 0;
+    arena->tail = NULL;
+    arena->current_chunk_end = NULL;
+}
+
+char* xb_arena_strdup(XbArena* arena, const char* src) {
+    if(!src) return NULL;
+    int n = strlen(src) + 1;
+    char* out = xb_arena_alloc(arena, n);
+    memcpy(out, src, n);
+    out[n] = '\0';
+    return out;
+}
+
+char* xb_arena_strndup(XbArena* arena, const char* src, size_t strsz) {
+    if(!src) return NULL;
+    /* strnlen isn't in the c99/posix standard even though it's widely supported
+     * easy enough to do it ourselves */
+    size_t n;
+    for(n=0; n<strsz; n++) {
+        if(src[n] == '\0') break;
+    }
+    char* out = xb_arena_alloc(arena, n+1);
+    memcpy(out, src, n);
+    out[n] = '\0';
+    return out;
+}
+
+XbArenaPtrArray* xb_arena_ptr_array_new(XbArena* arena) {
+    XbArenaPtrArray* self = xb_arena_alloc(arena, sizeof(XbArenaPtrArray));
+    *self = (XbArenaPtrArray){
+        .arena = arena,
+        .pointers = (void**) xb_arena_alloc(arena, 4 * sizeof(void*)),
+        .len = 0,
+        .cap = 4,
+    };
+    return self;
+}
+void xb_arena_ptr_array_add(XbArenaPtrArray* self, void* data) {
+    /* Grow array if necessary */
+    if(self->len+1 > self->cap) {
+        self->cap *= 2;
+        void** newptrs = xb_arena_alloc(self->arena, self->cap * sizeof(void*));
+        memcpy(newptrs, self->pointers, self->len * sizeof(void*));
+        self->pointers = newptrs;
+    }
+    /* Insert new pointer */
+    self->pointers[self->len] = data;
+    self->len++;
+}
+void xb_arena_ptr_array_remove(XbArenaPtrArray* self, void* data) {
+    for (guint i = 0; i < self->len; i++) {
```

Comment:
(Putting this comment here arbitrarily, but it applies to the whole file)

What’s the thread safety situation for the arena allocator? Can we guarantee that it (and its utilities, like `xb_arena_ptr_array_remove()`) will only ever be called from a single thread for a given `XbArena`?